### PR TITLE
Handle empty input in AddStringKeyStoreCommand

### LINKED ISSUE #39413: elasticsearch-keystore utility crashes with NPE when passed 0-byte closed STDIN and `-x`
**Elasticsearch version** (`bin/elasticsearch --version`):

~~~
Version: 8.0.0-SNAPSHOT, Build: default/tar/ed20d7c/2019-02-25T08:50:10.599708Z, JVM: 11.0.1
~~~

**Plugins installed**: none

**JVM version** (`java -version`):
~~~
java 11.0.1 2018-10-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.1+13-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)
~~~

**OS version** (`uname -a` if on a Unix-like system):

~~~
Darwin castrovel.local 16.7.0 Darwin Kernel Version 16.7.0: Sun Oct 28 22:30:19 PDT 2018; root:xnu-3789.73.27~1/RELEASE_X86_64 x86_64
~~~

**Description of the problem including expected versus actual behavior**:

When the Elasticsearch Keystore is invoked with `-x`, and the stdin stream is closed with zero bytes, the tool crashes with an NPE in its stacktrace.

**Steps to reproduce**:

~~~
╭─{ yaauie@castrovel:~/src/elastic/logstash/build/elasticsearch (✘ master) }
╰─● printf '' | bin/elasticsearch-keystore add foo -f -x
Exception in thread "main" java.lang.NullPointerException
	at org.elasticsearch.common.settings.AddStringKeyStoreCommand.execute(AddStringKeyStoreCommand.java:87)
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.MultiCommand.execute(MultiCommand.java:77)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.Command.main(Command.java:90)
	at org.elasticsearch.common.settings.KeyStoreCli.main(KeyStoreCli.java:40)
[error: 1]
~~~

### PR DESCRIPTION
**Pull request by jkakavas**
This change ensures that we do not make assumptions about the length
of the input that we can read from stdin and that we can handle empty 
strings. The behavior remains the same with regards to the fact that we
don't accept multi-line input 

Resolves: #39413

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java [hunk 1]
```diff
@@ -83,8 +85,19 @@ protected void execute(Terminal terminal, OptionSet options, Environment env) th
...
+                int charInt;
+                while ((charInt = stdinReader.read()) != -1) {
+                    if ((char) charInt == '\r' || (char) charInt == '\n') {
+                        break;
+                    }
+                    writer.write((char) charInt);
+                }
+                value = writer.toCharArray();
+            } catch (IOException e) {
+                throw new UserException(ExitCodes.DATA_ERROR, e.getMessage());
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java]: UserException is meant to be something clear for users which does not warrant a stack trace. Is the raw IOException message actually clear? What possible error scenarios exist?

Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java]: You're right, this was an oversight. IOException can be thrown by `read()` in the ByteBuffer, so I added a generic message and added the IOException as cause in the UserException

**Comment by jkakavas** [comment]
This is ready for another pass @rjernst , thanks

**Comment by jkakavas** [comment]
Ping @rjernst

File: server/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java [hunk 2]
```diff
@@ -83,8 +85,19 @@ protected void execute(Terminal terminal, OptionSet options, Environment env) th
...
+                int charInt;
+                while ((charInt = stdinReader.read()) != -1) {
+                    if ((char) charInt == '\r' || (char) charInt == '\n') {
+                        break;
+                    }
+                    writer.write((char) charInt);
+                }
+                value = writer.toCharArray();
+            } catch (IOException e) {
+                throw new UserException(ExitCodes.DATA_ERROR, "Error reading the value from stdin", e);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java]: This message is still opaque to a user. Remember UserException does *not* print a stacktrace. In this case, I think we should just let IOException bubble up and print the full stacktrace? What benefit is there in wrapping it? If there is a subclass of io errors that the user can directly do something about, then we should distinguish that and message appropriately with actions the user can take.

Comment (jkakavas) [reply server/src/main/java/org/elasticsearch/common/settings/AddStringKeyStoreCommand.java]: I got the idea to add the exceptions to the UserException from Jay's feedback in https://github.com/elastic/elasticsearch/pull/38498#discussion_r260391108 but I see what you're saying here. There is no subset of i/o errors that I know of where we could print something useful for the user, so I'll let the IOException be thrown

**Review by rjernst — APPROVED** [review]
lgtm

**Pull request merged**